Fixes for AUI in dark mode under macOS

vadz opened the pull request:
Replaces #1950 and should close [#18812](https://trac.wxwidgets.org/ticket/18812).

dkulp committed:
Use system colours in wxAuiGenericTabArt

Don't hardcode black and grey, which don't look right in all themes.

dkulp committed:
Update AUI art colours after system colours change

Move code initializing colours and colour-dependent bitmaps from ctor to
UpdateColoursFromSystem(), which means that now this is done not only
once, when creating the object, but also later if the system colours
change (e.g. due to turning on or off the dark mode).

dkulp committed:
Adapt the gradient used in AUI tab art in dark mode

Invert its direction and make it darker for better fit.

dkulp committed:
Improve dark mode test in wxAuiGenericTabArt

Use wxSystemSettings for this instead of manually checking the base
colour components which is both more verbose/difficult and less
reliable.

dkulp committed:
Simplify bitmap initialization in wxAuiDefaultDockArt

Reduce the number of __WXMAC__ checks by using helper wxColour variables
which have different values under Mac and non-Mac.

No real changes.

dkulp committed:
Use more appropriate colours for AUI dock art under Mac

Use system colours instead of hard-coding white.

This is still not ideal because ideally the same wxSYS_COLOUR_XXX
constants should be usable under all platforms, but better than just
using completely wrong colours when using dark mode under Mac.

vadz commented:
@dkulp and @mimi89999, your comments are welcome (those of everybody else are too, of course). TIA!

dkulp reviewed (changes requested):
This needs to also set teh topLightness to 110, not 170.   170 is way too light for dark mode.  It's a very light gray.

dkulp commented on src/aui/tabart.cpp:
```diff
@@ -284,6 +285,11 @@ void wxAuiGenericTabArt::DrawBackground(wxDC& dc,
 
+    if (wxSystemSettings::GetAppearance().IsUsingDarkBackground())
+    {
+        wxSwap(bottomLightness, topLightness);
```
This needs to set the topLightness to 110, not 170.   170 is way too light for dark mode.  It's a very light gray.

vadz replied:
Oh, I overlooked this, thanks for noticing, will fix.

mimi89999 commented:
Inactive colors in tabart are not OK on GTK:

![Capture d’écran du 2020-07-13 10-55-13](https://user-images.githubusercontent.com/8530546/87285108-69402880-c4f7-11ea-94a6-3f034c68854c.png)
![Capture d’écran du 2020-07-13 10-55-30](https://user-images.githubusercontent.com/8530546/87285117-6ba28280-c4f7-11ea-9ee9-918dcb13c115.png)

vadz commented:
Yes, the tab close bitmap is not right... I have no idea which colours should be used for it nor what should it be.

Nevertheless this can't be applied because it actually makes things _worse_ for GTK. What would be the minimal fix to at least avoid the regression?

mimi89999 commented:
I think that it would be the best to use native GTK3 tabs. There is already `tabartgtk`, but it needs to be ported to GTK3. As for the minimal fix I would leave the old `wxColour(128,128,128)` on GTK.

vadz commented:
Using hardcoded colour (`#808080`) with `wxSYS_COLOUR_CAPTIONTEXT` is bound to result in problems in some themes. So either we should also use hardcoded black for the latter under GTK or we need to find something better for this system colour.

Using native tabs would be best, of course, but this won't be done in time for 3.1.4.

mimi89999 commented:
The system color is correct, but AUI changes the brightness of the background color and it managed to get a color very close to `wxSYS_COLOUR_INACTIVECAPTION` for the background. Let's also shift the brightness of `wxSYS_COLOUR_INACTIVECAPTION` in tabs then.

> Using native tabs would be best, of course, but this won't be done in time for 3.1.4.

Let's try to get it. AUI got already a lot of improvements. Let's try to make it perfect for 3.1.4.

vadz commented:
> Let's also shift the brightness of `wxSYS_COLOUR_INACTIVECAPTION` in tabs then.

How do you do this?

> Let's try to get it. AUI got already a lot of improvements. Let's try to make it perfect for 3.1.4.

Making it perfect for 3.2.0 is a more important goal. Right now we really need to release 3.1.4 and it's very unlikely we'll manage it until then. I'd rather be realistic.

mimi89999 commented:
```
if (dark) {
    inactive = wxSystemSettings::GetColour(wxSYS_COLOUR_INACTIVECAPTION).ChangeLightness(130)
} else {
    inactive = wxSystemSettings::GetColour(wxSYS_COLOUR_INACTIVECAPTION).ChangeLightness(70)
}
```

vadz commented:
Thanks! I guess this should be done for GTK only, right? I.e. @dkulp would object to doing it like this for macOS?

dkulp commented:
With the "alpha" fix that is attached to the JIRA, the color would be wxSYS_COLOUR_INACTIVECAPTIONTEXT, not wxSYS_COLOUR_INACTIVECAPTION which this PR currently uses.   Does that change the GTK colors?

dkulp commented:
With the the new "patch3.txt" I attached to the TRAC issue which fixes the alpha handling on OSX and allows using the wxSYS_COLOUR_INACTIVECAPTIONTEXT, the GTK now looks like:
<img width="829" alt="Screen Shot 2020-07-13 at 8 58 39 AM" src="https://user-images.githubusercontent.com/91569/87307280-4a399a80-c4e7-11ea-95c2-1f42cf7cf7bc.png">

which looks OK.

dkulp committed:
Honour the colour alpha channel in wxAuiBitmapFromBits()

This fixes creating wrong bitmaps for system colours with alpha and
allows to use the correct wxSYS_COLOUR_INACTIVECAPTIONTEXT constant
in wxAuiGenericTabArt instead of a wrong but, due to the loss of alpha,
better looking wxSYS_COLOUR_INACTIVECAPTION which was used before.

dkulp committed:
Use previously Mac-specific colours for all platforms in wxAUI

Get rid of platform checks in wxAUI and use (the same) system colours
everywhere.

vadz committed:
Remove the now unused DISABLED_TEXT_COLOR constant

The previous commit removes all references to this hardcoded colour.

vadz commented:
This should be in sync with the latest patches from wxTrac now. @mimi89999 please let me know if you still see any problems/regressions with this version. TIA!

mimi89999 commented:
Now there is no difference when the button is hovered, but that should probably be fixed somewhere else.

mimi89999 commented:
![Capture d’écran du 2020-07-13 16-31-45](https://user-images.githubusercontent.com/8530546/87317016-bbe50900-c526-11ea-906b-3c9c88ba506c.png)

mimi89999 commented on src/aui/auibar.cpp:
```diff
@@ -366,11 +354,7 @@ void wxAuiGenericToolBarArt::DrawButton(
-#ifdef __WXMAC__
         dc.SetTextForeground(wxSystemSettings::GetColour(wxSYS_COLOUR_INACTIVECAPTIONTEXT));
-#else
-        dc.SetTextForeground(DISABLED_TEXT_COLOR);
```
Disabled is not the same as inactive. There is no way to get the disabled color. If `wxSYS_COLOUR_INACTIVECAPTIONTEXT` should be the color of disabled, then that color should be changed.

mimi89999 commented:
@vadz Could you split the changes into several PRs? I think it will be really easier to review.

vadz commented:
Sorry, I'm bad at "find 10 differences" game, could you please explain what exactly is wrong with the latest screenshot?

I don't really understand how could these changes have broken appearance change on hover... does anybody have any idea?

As for disabled text, here is how it's [done in `wxGrid`](https://github.com/wxWidgets/wxWidgets/blob/69e27acf0df03a4868c4012639428fb423a1c5f1/src/generic/grid.cpp#L265,L292). I'm afraid this might not look correctly in the dark mode neither, but we really should have a function for rendering disabled text anyhow.

> @vadz Could you split the changes into several PRs? I think it will be really easier to review.

Sadly I can't because there is no subset of commits that could be merged, i.e. don't result in some regressions under either GTK or macOS. These are not my changes, BTW (well, except for the trivial parts), I'm just converting patches posted to wxTrac to the commits here. FWIW I'd be really glad if anybody could come up with something reasonable and working on all platforms without me having to touch this at all...

mimi89999 commented:
> Sorry, I'm bad at "find 10 differences" game, could you please explain what exactly is wrong with the latest screenshot?

Of course. You can see that on https://trac.wxwidgets.org/attachment/ticket/18820/Capture%20d%E2%80%99%C3%A9cran%20du%202020-07-13%2014-08-26.png the label of the disabled button is grayed out. Now it isn't. It seems that the `wxSYS_COLOUR_INACTIVECAPTIONTEXT` color is inconsistent between platforms. Inactive doesn't mean disabled.

> As for disabled text, here is how it's done in wxGrid. I'm afraid this might not look correctly in the dark mode neither, but we really should have a function for rendering disabled text anyhow.

In the doc (https://docs.wxwidgets.org/trunk/settings_8h.html) I see that there is `wxSYS_COLOUR_GRAYTEXT` _Colour of greyed (disabled) text_ and that seems to be exactly what we need. I hope that it works as expected on macOS.

> Sadly I can't because there is no subset of commits that could be merged

I think that the change making button color change after system theme change should be OK now that color cache invalidation works well on GTK.

mimi89999 commented:
@dkulp Will `wxSYS_COLOUR_GRAYTEXT` work for macOS?

dkulp commented:
On OSX, wxSYS_COLOUR_INACTIVECAPTIONTEXT and wxSYS_COLOUR_GRAYTEXT are identical, so yes.

vadz committed:
Use wxSYS_COLOUR_GRAYTEXT for disabled AUI buttons labels

This is more appropriate than wxSYS_COLOUR_INACTIVECAPTIONTEXT which
just happened to be the same as wxSYS_COLOUR_GRAYTEXT under Mac, but
this is not true elsewhere.

vadz commented:
OK, thanks, pushed another commit changing this. There is still a problem with hover state, IIUC, but if it's not too bad, maybe we could merge this and fix it later. But I'd like to get confirmation from both of you that this PR works satisfactorily before merging it. TIA!

dkulp commented:
Looks good to me.    The hover state works fine on Linux, just not on OSX.   However, it has nothing to do with any of these commits as it's not working with the commit just before all of these either.

mimi89999 commented:
I found a really bad issue. On some very popular themes the caption is dark, but the rest of the window is white.

Yaru (Ubuntu default theme)
![Capture d’écran du 2020-07-13 21-32-41](https://user-images.githubusercontent.com/8530546/87345493-6c1b3780-c550-11ea-8c1c-087cc4d9eafb.png)

Pop theme (Pop!_OS default theme)
![Capture d’écran du 2020-07-13 21-34-30](https://user-images.githubusercontent.com/8530546/87345643-a7b60180-c550-11ea-82cf-87a5c37e451a.png)

Materia (a very popular material theme)
![Capture d’écran du 2020-07-13 21-35-52](https://user-images.githubusercontent.com/8530546/87345807-e64bbc00-c550-11ea-9494-19d8c9f6ca13.png)

Numix
![Capture d’écran du 2020-07-13 21-36-48](https://user-images.githubusercontent.com/8530546/87345996-388cdd00-c551-11ea-9e7f-290078b11d4e.png)

It seems that caption colors shouldn't be used at all outside of the caption. Those are some of the most popular themes. A also missed that in my https://github.com/wxWidgets/wxWidgets/pull/1791.

Also, there are already 10 commits in this PR. I think that they should be squashed.

mimi89999 commented:
For the buttons any of the `wxSYS_COLOUR_WINDOWTEXT `, `wxSYS_COLOUR_BTNTEXT` and `wxSYS_COLOUR_LISTBOXTEXT` would be OK. Maybe more colors. I only checked colors with `text` in them. For disabled buttons `wxSYS_COLOUR_GRAYTEXT` can be used and for the highlighted close let's use the 30 lightness shift.

vadz commented:
@mimi89999 If you can make a PR against this branch (vadz:aui-dark-mode) fixing the colours for these themes, it would be really welcome because I am completely lost with all these colours. Otherwise I don't even know what should we do -- apply the changes here only in dark mode? Only for macOS? Burn everything down and start afresh?

mimi89999 commented:
Now I'm waiting for feedback on those colors on macOS from @dkulp.

mimi89999 commented:
Also, `m_disabledCloseBmp` is completely misleading. Once this will be merged I will open a PR to fix naming.

vadz commented:
All of `wxSYS_COLOUR_WINDOWTEXT`, `wxSYS_COLOUR_BTNTEXT` and `wxSYS_COLOUR_LISTBOXTEXT` are the same as `wxSYS_COLOUR_CAPTIONTEXT` in wxOSX currently.

mimi89999 commented:
Ah. Should have checked. :laughing:. I propose to use `wxSYS_COLOUR_GRAYTEXT` for disabled and `wxSYS_COLOUR_WINDOWTEXT` for active. It should be OK then. I hope you will squash the commits at some point.

vadz commented:
@mimi89999 Could you please post a patch which works for you? I'm afraid I'm going to miss/mix up some colours if I do it myself. TIA!

Also, which commits do you want to squash exactly?

mimi89999 commented:
Maybe it will be best to make tabart color changes macOS only. Let's not keep them blocked. I will have more time to test on GTK.

vadz commented:
Sorry, I'm not sure what exactly do you propose, could you please explain? AFAICS this PR still can't be merged, so how should it be (minimally) modified to allow doing it?

mimi89999 commented:
```
    m_activeCloseBmp = wxAuiBitmapFromBits(close_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOWTEXT));
    if (dark) {
        m_disabledCloseBmp = wxAuiBitmapFromBits(close_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_INACTIVECAPTIONTEXT).ChangeLightness(130));
    } else {
        m_disabledCloseBmp = wxAuiBitmapFromBits(close_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_INACTIVECAPTIONTEXT).ChangeLightness(70));
    }
    m_activeLeftBmp = wxAuiBitmapFromBits(left_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOWTEXT));
    m_disabledLeftBmp = wxAuiBitmapFromBits(left_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_GRAYTEXT));
    m_activeRightBmp = wxAuiBitmapFromBits(right_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOWTEXT));
    m_disabledRightBmp = wxAuiBitmapFromBits(right_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_GRAYTEXT));
    m_activeWindowListBmp = wxAuiBitmapFromBits(list_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_WINDOWTEXT));
    m_disabledWindowListBmp = wxAuiBitmapFromBits(list_bits, 16, 16, wxSystemSettings::GetColour(wxSYS_COLOUR_GRAYTEXT));
```

mimi89999 committed:
Use WINDOW/GRAYTEXT system colours for AUI tab art

Those are more appropriate than the previously used CAPTIONTEXT and
INACTIVECAPTIONTEXT under Linux and don't change anything under Mac,
where they map to the same colours anyhow.

mimi89999 committed:
Adjust disabled closed bitmap colour to better fit Linux themes

Change the lightness of INACTIVECAPTIONTEXT to make it stand out better.

vadz commented:
Thanks, I've added your changes now, please let me know if you (or @dkulp) see anything wrong/worse with this PR than without it.

dkulp commented:
I'm OK with it.  It looks fine on osx.

One note:  on OSX the change from wxSYS_COLOUR_WINDOW to wxSYS_COLOUR_3DFACE for the base color does lighten everything up fairly significantly.   However, I'm not sure if one is "more correct" than the other.   I've attached a screen shot where the top is with the previous values, bottom is with all these changes (and active pane turned off).   There is definitely a noticeable difference, but I'm not sure if either is right or wrong.
 
<img width="1653" alt="Screen Shot 2020-07-14 at 10 34 02 AM" src="https://user-images.githubusercontent.com/91569/87438472-948d4b00-c5bd-11ea-83e2-9cd4d23406e5.png">

vadz commented:
Frankly, at this point, unless it's blatantly wrong/unusable I'd leave it like this and perhaps try to improve things later, so let's merge this, thanks for testing!

I'd just like to have one final confirmation from @mimi89999 that it looks acceptable in Yaru/Pop themes too before doing it.

dkulp commented:
One of my users and I went through all the apps installed on our Macs and about 75% of them are closer to the "new" colors so merging it as it is makes sense.

mimi89999 commented:
No, `wxSystemSettings::GetAppearance().IsUsingDarkBackground() doesn't work. Do we have something like https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/demos/gtk-demo/foreigndrawing.c#L129 (the `get_style` function) in wxWidgets internals?

vadz commented:
Why doesn't `IsUsingDarkBackground()` work? For which theme does it happen and what are `wxSYS_COLOUR_WINDOW` and `wxSYS_COLOUR_WINDOWTEXT` values for it (you can just run the drawing sample using this theme to see it)?

I don't understand how does the function you show would help us, i.e. how could we use it to check if we're using dark background or not?

mimi89999 commented:
It is working fine now. I think that I tested on a broken theme.

> I don't understand how does the function you show would help us, i.e. how could we use it to check if we're using dark background or not?

Not for that, but for discovering theme colors and in the future GTK3 port of `tabartgtk`.

vadz commented:
OK, just to confirm: the latest version looks fine in the common themes you've tested above (Yaru/Pop/Materia) and so can be merged?

mimi89999 commented:
Yes

mimi89999 commented:
A nice enhancement would be to replace all `dc.SetTextForeground(wxSystemSettings::GetColour(wxSYS_COLOUR_CAPTIONTEXT));` with `dc.SetTextForeground(wxSystemSettings::GetColour(wxSYS_COLOUR_BTNTEXT));` in auibar, but this is not something that was broken by this PR, so I don't consider it an issue with this PR.

AliKet commented:
Hi,

I thought this is a macOS only changes, but it appears that the changes also apply to GTK.
if so, i'm glad to test this PR and share the screenshots this night ?

mimi89999 commented:
Yes, this PR also affects GTK3. Actually, it affects all platforms (besides the tabart changes that won't affect Windows and GTK2). I spent a lot of time testing this, but I think that more tests are always better. There are other issues with AUI that I would like to fix before the release, but they will conflict with this PR. I think that it would be best to test once everything will be fixed.

vadz commented:
@AliKet your testing is still welcome, but I've merged this because it's already too long and I prefer to fix any new problems in separate commits/PRs.

Thanks again to @dkulp and @mimi89999 for your patches and testing!

AliKet commented:
Hi again,

So i've run the AUI sample under 40 themes on Ubuntu 18.04 using GTK3. and here is it with 4 of my favorit themes:

**Nexwaita-dark:**
![Nextwaita-dark-1 3](https://user-images.githubusercontent.com/7704771/87477767-90dee000-c620-11ea-9ba5-f1d8966bf548.png)

**Adwaita-Dark-Green:**
![Adwaita-Dark-Green](https://user-images.githubusercontent.com/7704771/87477984-f16e1d00-c620-11ea-9930-e355b62cb15a.png)

**Lounge-night:**
![Lounge-night](https://user-images.githubusercontent.com/7704771/87476797-e619f200-c61e-11ea-9e97-ab3d030f7618.png)

**Mojave-dark:**
![Mojave-dark](https://user-images.githubusercontent.com/7704771/87477668-68ef7c80-c620-11ea-8623-e90f58a3373d.png)

**IOS-4-2.0:**
![IOS-4-2 0](https://user-images.githubusercontent.com/7704771/87478218-4b6ee280-c621-11ea-94c8-dc350b89e8b8.png)

vadz commented:
OK, so Adwaita-Dark-Green is plain unusable while iOS-4-2.0 is barely usable (am I missing anything else?).

The problem with both is that tools text uses light colour when it clearly should be using a dark one on this background, while for the "Disabled" tool it's vice versa.

AliKet commented:
If you want, i can upload the rest as zipped file (it is about 2.3 MB), just tell me how ?

vadz commented:
Sorry, I'm not sure I really want this, I'm probably just going to glance over all these pictures. It would be more useful to summarize the problems you see with them, ideally in a new Trac ticket. TIA!

mimi89999 commented:
#1966 should fix this.

mimi89999 commented:
@aliket I think that the disabled color is as expected. Could you show this in gtk3 widget factory?

AliKet commented:
> @AliKet I think that the disabled color is as expected. Could you show this in gtk3 widget factory?

Yes, I think it's expected and I guess it's just a theme bug to report an incorrect disabled colour from their side. Ex.:

**Correct disabled colour (Windows10 theme):**
![Windows10](https://user-images.githubusercontent.com/7704771/87486854-2041be80-c634-11ea-98df-e418e9679608.png)

**Incorrect disabled colour (Windows-Vista theme):**
![Windows-Vista-1 0](https://user-images.githubusercontent.com/7704771/87486885-33548e80-c634-11ea-809b-3ed181653115.png)
